File: 백준/Silver/26169. 세 번 이내에 사과를 먹자/세 번 이내에 사과를 먹자.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
int a[10][10];
bool cache[10][10];
int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };
bool hasAns = false;

void dfs(int x, int y, int cnt, int rm)
{
    if (rm == 0)
    {
        if (cnt >= 2)
        {
            // cout << x << ' ' << y << ' ' << cnt << endl;
            hasAns = true;
        }
        return;
    }

    for (int i = 0; i < 4; i++)
    {
        int nx = x + dx[i], ny = y + dy[i];
        if (nx < 0 || ny < 0 || nx >= 5 || ny >= 5) continue;
        if (cache[nx][ny] || a[nx][ny] == -1) continue;
        cache[x][y] = true;
        if (a[nx][ny] == 1)
            dfs(nx, ny, cnt + 1, rm - 1);
        else
            dfs(nx, ny, cnt, rm - 1);
        cache[x][y] = false;
    }
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    for (int i = 0; i < 5; i++) for (int j = 0; j < 5; j++)
        cin >> a[i][j];

    int x, y; cin >> x >> y;
    dfs(x, y, 0, 3);

    if (hasAns) cout << 1;
    else cout << 0;
}

File: 백준/Silver/26169. 세 번 이내에 사과를 먹자/세 번 이내에 사과를 먹자.py
cache = [[False] * 5 for _ in range(5)]
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
def dfs(x, y, cnt, move, a):
    if move == 3: 
        if cnt >= 2:
            return True
        return False

    has_ans = False
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if 0 <= nx < 5 and 0 <= ny < 5 and not cache[nx][ny] and a[nx][ny] != -1:
            cache[x][y] = True
            if a[nx][ny] == 1:
                has_ans |= dfs(nx, ny, cnt + 1, move + 1, a)
            else:
                has_ans |= dfs(nx, ny, cnt, move + 1, a)
            cache[x][y] = False

    return has_ans

# 입력 받기
a = [list(map(int, input().split())) for _ in range(5)]
x, y = map(int, input().split())

# DFS 실행
has_ans = dfs(x, y, 0, 0, a)

# 결과 출력
print(1 if has_ans else 0)